### PR TITLE
.NET Core breaking changes

### LINKED ISSUE #11566: Document the bar for breaking changes
**Help us make content visible**

AFAIK we don’t have good documentation on our approach to breaking changes.

Most people have a very simplistic view on this matter which is usually one of two extremes: _only changes to contracts_ vs. _any changes that can break my code_. Neither is very helpful because it doesn't accurate capture how we approach this.

**Describe the new topic**

We have a few documents in CoreFx here:

* https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-changes.md
* https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md
* https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-definitions.md

I think a good topic would start with a description on my the extremes aren't useful and that it's a continuum, along the lines of the first document. The second doc can serve as a set of useful example although I don't think the docs need to be as exhaustive. Example probably suffice.

The last doc has some useful terms we might want to define in this topic as well.

TOC wise, I think it would be good somewhere here:

image

@rpetrusha mentioned he'd be interested in helping getting this written.

/cc @mairaw

### PR DESCRIPTION
**Pull request by rpetrusha**
## .NET Core breaking changes

Fixes #11566 

//cc @terrajobst

### CODE REVIEW

File: docs/core/compatibility/categories.md
Comment (mairaw) [inline docs/core/compatibility/categories.md]: not sure if you wanted bold or italic

```suggestion
Backwards compatibility refers to the ability of an existing consumer of an API to run against a new version while behaving in the same way. Both behavioral changes and changes in binary compatibility affect backwards compatibility. If a consumer is not able to run or behaves differently when running against the newer version of the API, the API is *backwards incompatible*.
```

File: docs/core/compatibility/categories.md
Comment (mairaw) [inline docs/core/compatibility/categories.md]: adding product suffix 

```suggestion
title: Breaking change categories - .NET Core
```

File: docs/core/compatibility/categories.md
Comment (mairaw) [inline docs/core/compatibility/categories.md]:
```suggestion
description: Learn about the ways in which breaking changes are categorized.
```

File: docs/core/compatibility/categories.md
Comment (mairaw) [inline docs/core/compatibility/categories.md]: should you add an intro saying what this article is about?

File: docs/core/compatibility/categories.md
Comment (mairaw) [inline docs/core/compatibility/categories.md]: should this be a numbered list instead?

File: docs/core/compatibility/categories.md
Comment (mairaw) [inline docs/core/compatibility/categories.md]:
```suggestion
A particular change can affect compatibility in a number of ways. These fall into five categories:
```

File: docs/core/compatibility/categories.md
Comment (mairaw) [inline docs/core/compatibility/categories.md]: nay?

File: docs/core/compatibility/categories.md
Comment (mairaw) [inline docs/core/compatibility/categories.md]:
```suggestion
```

File: docs/core/compatibility/index.md
Comment (mairaw) [inline docs/core/compatibility/index.md]:
```suggestion
title: Evaluate breaking changes - .NET Core
```

File: docs/toc.yml
Comment (mairaw) [inline docs/toc.yml]: I don't think you need to clarify it's .NET Core here in the TOC

```suggestion
    - name: Evaluate breaking changes
```

File: docs/core/compatibility/index.md
Comment (mairaw) [inline docs/core/compatibility/index.md]:
```suggestion
This article outlines the categories of compatibility changes (or breaking changes) and the way in which the .NET team evaluates changes in each of these categories. An understanding of the how the .NET team approaches possible breaking changes is particularly helpful for developers who are opening pull requests in the [dotnet/corefx](https://github.com/dotnet/corefx) GitHub repository that modify the behavior of existing APIs.
```

File: docs/core/compatibility/index.md
Comment (mairaw) [inline docs/core/compatibility/index.md]:
```suggestion
  - The C# compiler increasingly tries to optimize `callvirt` to a normal call whenever the target method is non-virtual and is probably not null (such as a method accessed through the [?. null propagation operator](./../csharp/language-reference/operators/member-access-operators.md#null-conditional-operators--and-)).
```

File: docs/core/compatibility/index.md
Comment (mairaw) [inline docs/core/compatibility/index.md]:
```suggestion
- **❌ Changing the name of an assembly**
```

File: docs/core/compatibility/index.md
Comment (mairaw) [inline docs/core/compatibility/index.md]:
```suggestion

```

File: docs/core/compatibility/index.md
Comment (mairaw) [inline docs/core/compatibility/index.md]:
```suggestion
  Because the new exception is a subclass of an existing exception, previous exception handling code continues to handle the exception. For example, in .NET Framework 4, culture creation and retrieval methods began to throw an <xref:System.Globalization.CultureNotFoundException> instead of an <xref:System.ArgumentException> if the culture could not be found. Because <xref:System.Globalization.CultureNotFoundException> derives from <xref:System.ArgumentException>, this is an acceptable change.
```

File: docs/core/compatibility/index.md
Comment (mairaw) [inline docs/core/compatibility/index.md]:
```suggestion
- **❌ Changing a synchronous API to asynchronous (and vice versa)**
```

File: docs/core/compatibility/index.md
Comment (mairaw) [inline docs/core/compatibility/index.md]:
```suggestion
  Unrecoverable exceptions should not be caught but instead should be handled by a high-level catch-all handler. Therefore, users are not expected to have code that catches these explicit exceptions. The unrecoverable exceptions are:
```

File: docs/core/compatibility/index.md
Comment (mairaw) [inline docs/core/compatibility/index.md]:
```suggestion
- **✔️ Expanding the visibility of a member that is not [virtual](../../csharp/language-reference/keywords/sealed.md)**
```

**Review by mairaw — CHANGES_REQUESTED** [review]
LGTM @rpetrusha. Left some comments for you to consider.
I'll do a new review after some of the changes I've requested are implemented.

**Comment by mairaw** [comment]
There are also some warnings on the build report that you should take a look.

File: docs/core/compatibility/categories.md
Comment (Thraka) [inline docs/core/compatibility/categories.md]: Don't need the `- .NET Core` on the title.

**Comment by rpetrusha** [comment]
This is ready for another review, @mairaw.

File: docs/core/compatibility/categories.md
Comment (mairaw) [reply docs/core/compatibility/categories.md]: Yes, it's better to have @Thraka - that's why I suggested to add it. It's part of the SEO guidelines to have a brand suffix.
https://review.docs.microsoft.com/en-us/new-hope/quality/catstestcases/seo-rule?branch=master#title-metadata---title-suffix

File: docs/core/compatibility/categories.md
Comment (mairaw) [inline docs/core/compatibility/categories.md]: can affect compatibility in A particular change -> it seems you forgot part of the sentence here?

File: docs/core/compatibility/categories.md
Comment (mairaw) [inline docs/core/compatibility/categories.md]: "behave the way as an older version" reads a bit funny

File: docs/core/compatibility/categories.md
Comment (mairaw) [inline docs/core/compatibility/categories.md]: wondering if it would be better to have these as H2s

File: docs/core/compatibility/index.md
Comment (mairaw) [inline docs/core/compatibility/index.md]: This anchor is broken

**Review by mairaw — APPROVED** [review]
Looks good @rpetrusha but left a few more suggestions for you before you merge

**Comment by rpetrusha** [comment]
Thanks for the thorough review once again, @mairaw!

File: docs/core/compatibility/categories.md
Comment (Thraka) [reply docs/core/compatibility/categories.md]: I'm confused then 😄 I thought this is automatically appended via the docfx file??

File: docs/core/compatibility/index.md
Comment (rpetrusha) [reply docs/core/compatibility/index.md]: I've tried to group acceptable, unacceptable, and questionable items together, which I think is the best order, and to cross-reference related issues.

File: docs/core/compatibility/categories.md
Comment (mairaw) [reply docs/core/compatibility/categories.md]: Depends on the section of the docs. Not for every folder.

**Pull request merged**
